src/engine/TPA.cc

```diff
@@ -1400,271 +1399,203 @@ class TransitionSystemNetworkManager {
-        if (vid == graph.getEntry() or vid == graph.getExit()) {
-            networkMap.insert({vid, NetworkNode()});
-            continue;
-        }
-        networkMap.insert({vid, NetworkNode()});
-        auto & node = networkMap.at(vid);
-        node.solver = mkSolver();
+        auto [it, _] = networkMap.insert({vid, NetworkNode()});
+        auto & node = it->second;
+        node.preSafe = logic.getTerm_false();
+        node.postSafe = logic.getTerm_false();
         node.blocked_children = 0;
+        if (vid == graph.getEntry() or vid == graph.getExit()) { continue; }
+        node.solver = mkSolver();
         TransitionSystem ts = constructTransitionSystemFor(vid);
         node.solver->resetTransitionSystem(ts);
-        node.trulySafe = logic.getTerm_false();
-    }
-    TimeMachine tm(logic);
-    for (EId eid : adjacencyRepresentation.getOutgoingEdgesFor(graph.getEntry())) {
-        auto target = graph.getTarget(eid);
-        getNode(target).solver->resetInitialStates(logic.getTerm_true());
-    }
-    for (EId eid : adjacencyRepresentation.getIncomingEdgesFor(graph.getExit())) {
-        auto source = graph.getSource(eid);
-        networkMap.at(source).solver->updateQueryStates(logic.getTerm_true());
     }
     // Connect the network
+    auto isSelfLoop = [this](EId eid) { return graph.getSource(eid) == graph.getTarget(eid); };
+    // TODO: no need for reverse post order, better to process outgoing edges
     for (auto vid : reversePostOrder(graph, adjacencyRepresentation)) {
-        if (vid != graph.getExit()) {
-            auto outGoingEdges = adjacencyRepresentation.getOutgoingEdgesFor(vid);
-            outGoingEdges.erase(
-                std::remove_if(outGoingEdges.begin(), outGoingEdges.end(),
-                               [this](EId eid) { return graph.getSource(eid) == graph.getTarget(eid); }),
-                outGoingEdges.end());
-            for (auto edge : outGoingEdges) {
-                getNode(graph.getTarget(edge)).parents.push(edge);
-            }
-        }
         if (vid != graph.getEntry()) {
             auto incomingEdges = adjacencyRepresentation.getIncomingEdgesFor(vid);
-            incomingEdges.erase(
-                std::remove_if(incomingEdges.begin(), incomingEdges.end(),
-                               [this](EId eid) { return graph.getSource(eid) == graph.getTarget(eid); }),
-                incomingEdges.end());
             for (auto edge : incomingEdges) {
-                getNode(graph.getSource(edge)).children.push(edge);
+                if (isSelfLoop(edge)) { continue; }
+                getNode(graph.getSource(edge)).children.push_back(edge);
+                getNode(graph.getSource(edge)).blockedReason.push_back(PTRef_Undef);
             }
         }
     }
 }
 
+namespace {
+
+enum class NodeState { PRE, POST };
+struct Entry {
+    NodeState state;
+    SymRef node;
+    EId incomingEdge; // Edge used to get into this node, valid only for PRE state
```

Comment:
```suggestion
    std::optional<EId> incomingEdge; // Edge used to get into this node, valid only for PRE state
```
Maybe it's better to have it as an optional parameter?

Reply:
Indeed, it makes sense to use `optional` here. Changed!